🌟UI testing - Submit an enquiry form

amankumarrr opened the pull request:
- Related #1448  

Scheduled the UI-Testing to submit enquiry form on every Monday at 7 AM.

amankumarrr committed:
UI Testing - adding test cases

amankumarrr committed:
Running example for create lead

amankumarrr committed:
Refactoring the code

amankumarrr committed:
Adding extension for playwright

amankumarrr committed:
disabling prettier log error

amankumarrr committed:
Adding env variable for target_URL for testing

amankumarrr committed:
Adding key reference in PR and main Slot's config

amankumarrr committed:
Merge branch 'main' into ui-testing

amankumarrr committed:
Updating yarn.lock conflict

amankumarrr committed:
Removing test-examples from the project

amankumarrr committed:
Flow - Adding enquiry flow to run the tests

amankumarrr committed:
Updating the env variable

amankumarrr committed:
Adding trailing double quote

amankumarrr committed:
Updating secret with env variables for test run

amankumarrr committed:
Adding mask to the secret

amankumarrr committed:
Adding workflow_dispatch variable

amankumarrr committed:
Refactoring code

amankumarrr committed:
Changing the retention day

amankumarrr committed:
Removing the pull request branches

wicksipedia commented on .github/.env:
```diff
@@ -8,3 +8,4 @@ KEY_VAULT=kv-sswwebsite-9eb3
 APP_INSIGHT_PROD=appInsight-sswwebsite-9eb3
 APP_INSIGHT_DEV=appInsight-sswwebsite-dev-9eb3
 NEXT_PUBLIC_CHATBASE_BOT_ID=gnbZXuG_5tgI_R3yzKcOn
+TARGET_URL="https://app-sswwebsite-9eb3-pr-1464.azurewebsites.net/consulting/angular"
```
I think the URL host should be in the weekly workflow i.e. https://ssw.com.au
The test should have `/consulting/angular`

amankumarrr replied:
done - I replaced it with HOST_URL and added the route in the test itself

wicksipedia commented on .github/workflows/weekly-ui-tests.yml:
```diff
@@ -0,0 +1,78 @@
+  #  workflow_dispatch:
+
+  schedule:
+    # Monday at 7 AM UTC - https://cron.help/#0_12_*_*_MON
```
should run before business hours on monday
add sydney time to the comment

amankumarrr replied:
done

wicksipedia commented on .github/workflows/weekly-ui-tests.yml:
```diff
@@ -0,0 +1,78 @@
+    inputs:
+      TARGET_URL:
+        type: string
+        description: "Target URL"
```
add a default value `https://ssw.com.au/consulting/angular`

amankumarrr replied:
Not done.

Since we are using HOST_URL, this will remain empty and it's not a required field so by default it will run https://ssw.com.au/consulting/angular

wicksipedia commented on .github/workflows/weekly-ui-tests.yml:
```diff
@@ -0,0 +1,78 @@
+        with:
+          name: playwright-report
+          path: playwright-report/
+          retention-days: 1
```
```suggestion
          retention-days: 14
```
2 weeks of reports should be enough

amankumarrr replied:
done

wicksipedia commented on .github/workflows/weekly-ui-tests.yml:
```diff
@@ -0,0 +1,78 @@
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 18
```
we should have an .nvmrc file in the repo - best to use that to remain consistent

amankumarrr replied:
done

wicksipedia commented on .gitignore:
```diff
@@ -44,4 +44,6 @@ yarn-error.log*
 # Partytown lib files
 /public/~partytown
-
+/test-results/
```
```suggestion

# Playwright output
/test-results/
```

amankumarrr replied:
done

wicksipedia commented on components/bookingForm/bookingForm.tsx:
```diff
@@ -177,18 +178,21 @@ export const BookingForm = (props) => {
                     label={ACTIVE_INPUT.FullName}
                     type="text"
                     {...getCommonFieldProps(FORM_INPUT.FullName)}
+                    id={FORM_INPUT_IDS.FullName}
```
lookup via name prop to avoid extra code

wicksipedia commented on components/button/utilityButton.tsx:
```diff
@@ -27,6 +27,7 @@ export const UtilityButton = ({
   const baseComponent = (
     <Button
       ripple
+      id="bookingButton"
```
this looks like a generic button - I wouldn't set a fixed id on this component

wicksipedia commented on components/util/constants.tsx:
```diff
@@ -88,6 +88,21 @@ export const FORM_INPUT = {
+  ReferredEmail: "bookingForm-referredEmail",
+  ReferralSource: "bookingForm-referralSource",
+};
+
```
these constants look like they are specific to the booking form - it should be co-located with the booking form

wicksipedia commented on components/util/constants.tsx:
```diff
@@ -162,3 +177,6 @@ export const ReferralSourceList = Object.keys(ReferralSource).map((key) => {
   };
 });
+
+export const SUCCESS_MESSAGE =
```
this is booking form specific - should be with the booking form

amankumarrr replied:
done

wicksipedia commented on pages/api/create-lead.ts:
```diff
@@ -20,11 +30,18 @@ export default async function handler(
+      const key_matched =
+        secretNote === process.env.SECRET_KEY_TO_BYPASS_RECAPTCHA;
+
+      if (key_matched) req.body.Note = TEST_NOTE;
```
if recaptcha is bypassed - this method only needs to scrub the secret from the message

amankumarrr replied:
done

wicksipedia commented on pages/api/create-lead.ts:
```diff
@@ -12,6 +12,16 @@ import { PA_FLOW } from "../../services/server/power-automate-flow";
 import { CustomError } from "../../services/server/customError";
 
+const TEST_NOTE =
+  "<br><br>Hi Account Managers,<br><br> This is a weekly test email to verify that the create lead flow is working.";
```
this should be in the ui test

amankumarrr replied:
done

wicksipedia commented on pages/api/create-lead.ts:
```diff
@@ -12,6 +12,16 @@ import { PA_FLOW } from "../../services/server/power-automate-flow";
+const TEST_NOTE =
+  "<br><br>Hi Account Managers,<br><br> This is a weekly test email to verify that the create lead flow is working.";
+
+const Test_RECAPATCHA_VALIDATION = {
```
```suggestion
const RECAPATCHA_VALIDATION_SUCCESS_RESULT = {
```

amankumarrr replied:
done

wicksipedia commented on ui-tests/forms.spec.ts:
```diff
@@ -0,0 +1,39 @@
+  await page.fill("#" + FORM_INPUT_IDS.Phone, TEST_PAYLOAD.Phone);
+  await page.selectOption("#" + FORM_INPUT_IDS.Location, TEST_PAYLOAD.Country);
+  await page.fill("#" + FORM_INPUT_IDS.Company, TEST_PAYLOAD.Company);
+  await page.selectOption("#" + FORM_INPUT_IDS.States, TEST_PAYLOAD.State);
```
since we are removing the ids earlier on - these selector will need to change to look for the `name` attr

amankumarrr replied:
done

amankumarrr committed:
Matt's Feedback

amankumarrr committed:
Adding doc for running it on local

amankumarrr committed:
Updating host url to prod

amankumarrr committed:
Renamed the variable name

amankumarrr committed:
Merge branch 'main' into ui-testing

amankumarrr committed:
Update create-lead

amankumarrr committed:
removing Id from util button

amankumarrr committed:
Merge branch 'main' into ui-testing

amankumarrr committed:
Matt's feedback 2

wicksipedia commented on pages/api/create-lead.ts:
```diff
@@ -12,16 +12,39 @@ import { invokePowerAutomateFlow } from "../../services/server/power-automate-fl
       const { Recaptcha } = req.body;
+
+      const Note = req.body.Note;
+      const key_matched = Note.includes(
```
```suggestion
      // Note: bypassing recaptcha is intended for weekly testing the lead capture form only
      const key_matched = Note.includes(
```

wicksipedia commented on pages/api/create-lead.ts:
```diff
@@ -12,16 +12,39 @@ import { invokePowerAutomateFlow } from "../../services/server/power-automate-fl
 ) {
   try {
     if (req.method === "POST") {
       const { Recaptcha } = req.body;
```
```suggestion
      // this is the code to validate with the recaptcha service
      const { Recaptcha } = req.body;
```

wicksipedia commented on components/bookingButton/bookingButton.tsx:
```diff
@@ -35,8 +38,8 @@ export const BookingButton = ({ data }) => {
-      <div className="text-left">
-        Form submitted. We'll be in contact as soon as possible.
+      <div id="success-toaster" className="text-left">
+        {SUCCESS_MESSAGE}
```
```suggestion
        Form submitted. We'll be in contact as soon as possible.
```

wicksipedia commented on components/bookingButton/bookingButton.tsx:
```diff
@@ -20,6 +20,9 @@ export interface BookingButtonProps {
 
+const SUCCESS_MESSAGE =
+  "Form submitted. We'll be in contact as soon as possible.";
+
```
```suggestion
```

wicksipedia commented on ui-tests/forms.spec.ts:
```diff
@@ -0,0 +1,75 @@
+import { expect, test } from "@playwright/test";
+import { FORM_INPUT } from "../components/util/constants";
```
i'd like to remove this

wicksipedia commented on ui-tests/forms.spec.ts:
I think this file should be renamed so that it is clear that it is testing the create lead form

amankumarrr committed:
Update pages/api/create-lead.ts

Co-authored-by: Matt Wicks [SSW] <wicksipedia@users.noreply.github.com>

amankumarrr committed:
Update pages/api/create-lead.ts

Co-authored-by: Matt Wicks [SSW] <wicksipedia@users.noreply.github.com>

amankumarrr committed:
Update components/bookingButton/bookingButton.tsx

Co-authored-by: Matt Wicks [SSW] <wicksipedia@users.noreply.github.com>

amankumarrr committed:
Update components/bookingButton/bookingButton.tsx

Co-authored-by: Matt Wicks [SSW] <wicksipedia@users.noreply.github.com>

amankumarrr committed:
Updated filename to make it more descriptive

wicksipedia commented on .github/workflows/weekly-ui-tests.yml:
```diff
@@ -0,0 +1,79 @@
+  #  workflow_dispatch:
+
+  schedule:
+    # Monday at 9 PM UTC ~ 7 AM AEST - https://cron.help/#0_21_*_*_SUN
```
```suggestion
    # Monday at 9 PM UTC ~> 7 AM AEST - https://cron.help/#0_21_*_*_SUN
```

wicksipedia committed:
Update .github/workflows/weekly-ui-tests.yml

wicksipedia committed:
♻️ general cleanup
- ui tests - renamed file to match component being tests
- ui tests - dry'd up code, preferred placeholders over attribute selectors where possible
- changed name of secret to bypass recaptcha (prefixed with recaptcha)
- playwright config - base url for tests should be set at this level
- readme - updated to run the ui tests the same way as the github action
- env file - updated example as it was out of sync with real settings

wicksipedia committed:
lint fixes
